Setup logging library

nhuray opened the pull request:
Hi @jekkel 

This is a PR related to issue #189 to replace the logs generated using `print` by logs managed with python `logging` [library](https://docs.python.org/3/library/logging.html).

I removed all the `${timestamp()}`in the logs because the library should generate a timestamp by default.

I defined a  `LOG_LEVEL`  environment variable to control the log level globally.

I put the most verbose logs in `debug`.


Nicolas

schans commented on src/helpers.py:
```diff
@@ -5,11 +5,18 @@
+
+# Initialize Logger
+logger = logging.getLogger(__name__)
+logger.setLevel(log_level)
```
Doesn't the setLevel method require a numeric value? 
Something like:
numeric_level = getattr(logging, log_level.upper(), None)
# ignore invalid log level setting
    if isinstance(numeric_level, int):
        logger.setLevel(numeric_level)

https://docs.python.org/3/library/logging.html#levels

nhuray replied:
@schans Looking at the code for `setLevel` it looks `int` and `str`are supported: 

```python
def setLevel(self, level):
        """
        Set the logging level of this logger.  level must be an int or a str.
        """
        self.level = _checkLevel(level)
        self.manager._clear_cache()

def _checkLevel(level):
    if isinstance(level, int):
        rv = level
    elif str(level) == level:
        if level not in _nameToLevel:
            raise ValueError("Unknown level: %r" % level)
        rv = _nameToLevel[level]
    else:
        raise TypeError("Level not an integer or a valid string: %r" % level)
    return rv
`
``

schans replied:
Today I learned ;-) Even better. We could consider some error handling though.

schans replied:
May I suggest something like:

```python
# Initialize Logger
logger = logging.getLogger(__name__)

# Set log level
try:
    log_level = os.getenv("LOG_LEVEL", logging.INFO)
    log_level = log_level.upper() if isinstance(log_level, str) else log_level
    logger.setLevel(log_level)
except (ValueError, TypeError) as e:
   logger.error(f"Unable to set log level: {e}")
```

nhuray replied:
Thanks for the suggestion @schans 👍

jekkel commented on src/resources.py:
```diff
@@ -17,6 +18,12 @@
 from helpers import request, write_data_to_file, remove_file, timestamp, unique_filename, CONTENT_TYPE_TEXT, \
     CONTENT_TYPE_BASE64_BINARY, execute, WATCH_SERVER_TIMEOUT, WATCH_CLIENT_TIMEOUT
 
+log_level = os.getenv("LOG_LEVEL", logging.INFO)
```
I don't like that the environment variable is read multiple times. Can't we centralize this and provide a parsed/validated log level in a central place which gets used by all instantiated loggers then?

nhuray replied:
We might centralize in `helpers.py` considering `resources.py`and `sidecar.py` imports `helpers.py`. 

Or we can create a dedicated `logger.py` to initialize the logger and define error handling as suggested by @schans.

Let me know what you prefer !

jekkel replied:
I'd go for `helpers.py` :)

nhuray replied:
Done !

jekkel reviewed (changes requested):
Looks good, thanks for the contribution! I can only see a minor style issue.

nhuray commented on src/sidecar.py:
```diff
@@ -1,12 +1,11 @@
 from kubernetes.config.kube_config import KUBE_CONFIG_DEFAULT_LOCATION
 from requests.packages.urllib3.util.retry import Retry
 
-from helpers import timestamp, REQ_RETRY_TOTAL, REQ_RETRY_CONNECT, REQ_RETRY_READ, REQ_RETRY_BACKOFF_FACTOR
```
I removed unused import here

nhuray commented on src/resources.py:
```diff
@@ -14,7 +14,7 @@
 from urllib3.exceptions import MaxRetryError
 from urllib3.exceptions import ProtocolError
 
-from helpers import request, write_data_to_file, remove_file, timestamp, unique_filename, CONTENT_TYPE_TEXT, \
```
I removed unused import here

nhuray commented:
@jekkel Looking at the logs stored as Github actions [artifacts](https://github.com/kiwigrid/k8s-sidecar/suites/6651032201/artifacts/251339378), I just saw the logs does not have the format expected.

Before:
```
[2022-05-25 08:33:30] No folder annotation was provided, defaulting to k8s-sidecar-target-directory
```

After:
```
No folder annotation was provided, defaulting to k8s-sidecar-target-directory
```

It looks by default the formatter only display the [message](https://docs.python.org/3/library/logging.html#formatter-objects). So should we define a default log format for backward compatibility and define a `LOG_FORMAT` environment variable to set a format if needed ? 

Talking about backward compatibility I think we should not set the log level to `INFO` if the `LOG_LEVEL` variable is not set cause we gonna change the current behaviour (log everything).

What do you think ?

schans commented:
> @jekkel Looking at the logs stored as Github actions [artifacts](https://github.com/kiwigrid/k8s-sidecar/suites/6651032201/artifacts/251339378), I just saw the logs does not have the format expected.
> 
> Before:
> 
> ```
> [2022-05-25 08:33:30] No folder annotation was provided, defaulting to k8s-sidecar-target-directory
> ```
> 
> After:
> 
> ```
> No folder annotation was provided, defaulting to k8s-sidecar-target-directory
> ```
> 
> It looks by default the formatter only display the [message](https://docs.python.org/3/library/logging.html#formatter-objects). So should we define a default log format for backward compatibility and define a `LOG_FORMAT` environment variable to set a format if needed ?

Something this might do the trick:
```python
logging.basicConfig(format='%(asctime)s %(levelname)s %(message)s')
```
We could make it configurable. For us that wouldn't bring that much extra value. Time and level are useful. 

We would prefer logging in json format following ECS standard, but that's an easy improvement after this PR is in.

 > Talking about backward compatibility I think we should not set the log level to `INFO` if the `LOG_LEVEL` variable is not set cause we gonna change the current behaviour (log everything).
> 
> What do you think ?

INFO or WARING default level makes sense to me. Log everything (debug level) is a bit too much and to some extend the direct reason for this PR ;-) But since it's configurable it wouldn't make that much of a difference.

schans commented on src/helpers.py:
```diff
@@ -35,6 +36,29 @@
+        return log
+    except (ValueError, TypeError) as e:
+        log.error(f"Unable to set log level: {e}")
+        exit(1)
```
I would prefer not to exit here. It just falls back to INFO (default) level and can run just fine.

nhuray replied:
make sense. In this case I would define the log as a warning:

```
log.warning("Unable to set log level: {e}. Define default level to INFO")
```

schans commented:
Or if we want to have timestamps in RFC3339:

```python
logging.basicConfig(format="%(asctime)s %(levelname)s %(message)s", datefmt='%Y-%m-%dT%H:%M:%S%z')
```

That would result in:
```
>>> logging.error('this is an error log message')
2022-05-25T14:17:20+0200 ERROR this is an error log message
```

nhuray commented:
@schans @jekkel 

Finally I added a `logger.py`  in order to centralize the logging capabilities.

I implemented 2 formatters which I guess should be good enough for now:
 - `LOGFMT`: Format log using [logfmt](https://brandur.org/logfmt) format
 - `JSON`: Format log in json

This is an example of the log generated:

**Logfmt**
```
time=2022-05-25T23:15:45.581453 level=WARNING msg="No folder annotation was provided, defaulting to k8s-sidecar-target-directory"
```

**Json**
```
{"time": "2022-05-25T23:24:20.836608", "level": "WARNING", "msg": "No folder annotation was provided, defaulting to k8s-sidecar-target-directory"}
```

@schans I did not implemented the ECS format cause it looks quite verbose to me for a so small project.

jekkel commented on src/logger.py:
```diff
@@ -0,0 +1,108 @@
+    :return: a logger instance
+    """
+    log = logging.getLogger(name)
+    ch = logging.StreamHandler()
```
I think configuring the root logger globally instead of each created loggers individually might improve readability a bit, wdyt?

nhuray replied:
agree

nhuray replied:
I just configured a single logger named `k8s-sidecar` that I'm using everywhere.

jekkel commented on src/logger.py:
```diff
@@ -0,0 +1,108 @@
+        return " ".join(log_format_msg)
+
+
+class JsonFormatter(BaseFormatter):
```
Shouldn't we just use `python-json-logger` (https://pypi.org/project/python-json-logger/) for JSON logging instead of implementing it here?

nhuray replied:
Might be a good option as well. I was just reticent to add dependencies to the project to keep it lean.

But we might consider those library are sufficiently small to not harm ! 

So up to you @jekkel ! Just le me know what you prefer

nhuray replied:
I added the dependency you suggested to log in Json, along another one to format in LogFmt. 

If we don't need the second one let me know I can remove it !

jekkel commented on src/logger.py:
```diff
@@ -0,0 +1,108 @@
+        return fields
+
+
+class LogfmtFormatter(BaseFormatter):
```
I think configuring the [default log formatter](https://docs.python.org/3/library/logging.html#formatter-objects) with suitable format strings should be good enough. I don't see why we need a custom formatter.

nhuray replied:
Here is a custom formatter to use LogFmt fomat (key and value). 

We can also use a format string like `time=%(asctime)s level=%(levelname)s msg=%(message)s` .

I centralize the log fields  in `BaseFormatter` in order to use the same for JsonFormatter and LogFmtFormatter but I can also revisit that to simplify things.

jekkel commented on README.md:
```diff
@@ -81,3 +81,5 @@ If the filename ends with `.url` suffix, the content will be processed as a URL
 | `WATCH_CLIENT_TIMEOUT`     | If you have a network outage dropping all packets with no RST/FIN, this is how many seconds your client waits on watches before realizing & dropping the connection. You can keep this number low. ([#85](https://github.com/kiwigrid/k8s-sidecar/issues/85))                                                                       | false    | `66`                            ...
 | `IGNORE_ALREADY_PROCESSED` | Ignore already processed resource version. Avoid numerous checks on same unchanged resource. req kubernetes api >= v1.19                                                                                                                                                                                                            | false    | `false`                         ...
+| `LOG_LEVEL`                | Set the logging level. (DEBUG, INFO, WARN, ERROR, CRITICAL)                                                                                                                                                                                                                                                                         | false    | `INFO`                          ...
+| `LOG_FORMAT`               | Set a log format. (JSON or LOGFMT)                                                                                                                                                                                                                                                                                                  | false    | `LOGFMT`                        ...
```
How about making this config then " ... `JSON` or a format string for the [default formatter](https://docs.python.org/3/library/logging.html#logging.Formatter)" ?

nhuray replied:
You mean make Json the default format ? 

Absolutely, I personally like LogFmt because it's more human readable and les verbose that Json. But Json is probably more versatile for parsing by log aggregator (Loki, ELK ...)

nhuray replied:
I changed to have `JSON` has the default

jekkel commented on src/helpers.py:
```diff
@@ -65,7 +70,7 @@ def write_data_to_file(folder, filename, data, data_type=CONTENT_TYPE_TEXT):
 
         if sha256_hash_new.hexdigest() == sha256_hash_cur.hexdigest():
-            print(f"{timestamp()} Contents of {filename} haven't changed. Not overwriting existing file")
+            logger.warning(f"Contents of {filename} haven't changed. Not overwriting existing file")
```
-> debug level, updates of config/map secrets not touching all entries can expected and should not create `WARNING` logs.

nhuray replied:
agree

nhuray commented on src/resources.py:
```diff
@@ -92,12 +96,12 @@ def list_resources(label, label_value, target_folder, request_url, request_metho
         if ignore_already_processed:
             if _resources_version_map.get(metadata.namespace + metadata.name) == metadata.resource_version:
-                # print(f"{timestamp()} Ignoring {resource} {metadata.namespace}/{metadata.name}")
+                # logger.debug(f"{timestamp()} Ignoring {resource} {metadata.namespace}/{metadata.name}")
```
good point

schans commented on src/logger.py:
```diff
@@ -0,0 +1,108 @@
+        fields = dict()
+        fields['time'] = self.formatTime(record)
+        fields['level'] = record.levelname
+        fields['msg'] = record.getMessage()
```
We could chose to align these fields with the ECS standard. That would entail:
time -> @timestamp
level -> log.level
msg -> message

For an example see:
https://www.elastic.co/guide/en/ecs-logging/python/current/installation.html
Full ref:
https://www.elastic.co/guide/en/ecs/current/ecs-field-reference.html

nhuray replied:
@schans I understand the need to have a uniform format but I'm not sure ECS is widely used in community. 

So why use more ECS over [Gelf](https://docs.graylog.org/docs/gelf) or [Json Well Format](https://github.com/JSONWellLogFormat/JSONWellLogFormat) ?

From my perspective what's is important is to have a Json format cause all log collectors / aggregators supports JSON. 

I implemented LogFmt as well, cause it's a format widely used in Go community, but I let the maintainers deciding to remove it or not !

schans replied:
I guess it's a matter of perspective on what you are used to ;-)

Looking at the Gelf format I wouldn't be a fan of numeric log levels and unix timestamps. 

For the Json Well Format, I'm not sure if that's the one you wanted to link to. It seems specific to (drilling) well measurements in the oil and gas industry.

I don't mind that much about the field names. I think most log aggregation software out there can easily rename and transform. Imho timestamp is a bit more specific than time. If we use msg instead of message you could also opt for ts instead of timestamp. I have seen that plenty as well. But if you prefer to keep the names as they are now that is perfectly fine as well.

nhuray replied:
My reasoning about using `time`, `level` and `msg` is to have fields semantically comprehensible for humans and sufficiently small to limit the size it takes to store the fields in a reversed index (eg. Elasticsearch, Loki...).

That format is what used the golang lib [Logrus](https://github.com/sirupsen/logrus). I saw a lot of Kubernetes projects using `logrus` for logging so considering that project is also for running in Kubernetes I adopt the same log format.

schans replied:
Aha. Yes, that makes sense.

schans commented on src/logger.py:
```diff
@@ -0,0 +1,108 @@
+        """
+        ct = self.converter(record.created)
+        ct_with_ms = time.mktime(ct) + (record.msecs / 1000)
+        ct_full = datetime.datetime.fromtimestamp(ct_with_ms)
```
Could we add the timezone info here? That would reduce the confusion when trying to correlate different log events. 
 Is that info available in the record? Or is the timestamp always in the system timezone? In that case we could use something link:

```python
ct_full = datetime.datetime.fromtimestamp(ct_with_ms, datetime.datetime.utcnow().astimezone().tzinfo)
```

nhuray replied:
@schans I just made a change to expose a `LOG_TZ` variable to choose time formatted in local timezone or UTC timezone. 

The `time` fields is formatted in ISO 8601 format (e.g `2022-05-28T14:24:04.037058+00:00`)

schans replied:
Thanks.

From my local tests and what I understand from the docs, the isoformat() function will only add the TZ info if that is present in the timestamp (the +00:00 part), but maybe that's something local to my setup.

nhuray replied:
I'm using the record `created` (POSIX timestamp) and format it like that:
```
datetime.fromtimestamp(record.created, log_tz).isoformat()
```
So I format to IS0 8601 format a `datetime` with `tzinfo`.

From my tests it gives:

**UTC TZ**
```
{"time": "2022-05-29T11:22:11.377220+00:00", "level": "INFO", "msg": "Starting collector"}
```


**LOCAL TZ**
```
{"time": "2022-05-29T07:21:56.590056-04:00", "level": "INFO", "msg": "Starting collector"}
```

schans replied:
Nice! Perfect.

otchilin commented on src/sidecar.py:
```diff
@@ -87,17 +91,17 @@ def main():
 
             if len(v_major) and len(v_minor) and (int(v_major) > 1 or (int(v_major) == 1 and int(v_minor) >= 19)):
-                print(f"{timestamp()} Ignore already processed resource version will be enabled.")
+                logger.debug("Ignore already processed resource version will be enabled.")
```
Hi !
Perhaps you may log these ones in 'info' level and 'warning' for the api compatibility issue instead of debug.

Thanks.
Regards.

nhuray replied:
Good catch ! Thanks

nhuray commented:
@jekkel Is there something blocking to merge that PR now ?

nhuray commented:
> To me everything looks good to go.
> 
> Thanks @nhuray for the work and patience dealing with all my comments!

Thanks for you review
